addons/web/static/src/js/widgets/switch_company_menu.js

```diff
@@ -98,19 +106,32 @@ var SwitchCompanyMenu = Widget.extend({
         ev.preventDefault();
         ev.stopPropagation();
         var dropdownItem = $(ev.currentTarget).parent();
-        var companyID = dropdownItem.data('company-id');
-        var allowed_company_ids = this.allowed_company_ids;
-        var current_company_id = allowed_company_ids[0];
+        const allowedCompanyIds = this.allowed_company_ids;
+        const currentCompanyId = allowedCompanyIds[0];
         if (dropdownItem.find('.fa-square-o').length) {
-            allowed_company_ids.push(companyID);
             dropdownItem.find('.fa-square-o').removeClass('fa-square-o').addClass('fa-check-square');
             $(ev.currentTarget).attr('aria-checked', 'true');
         } else {
-            allowed_company_ids.splice(allowed_company_ids.indexOf(companyID), 1);
             dropdownItem.find('.fa-check-square').addClass('fa-square-o').removeClass('fa-check-square');
             $(ev.currentTarget).attr('aria-checked', 'false');
         }
-        session.setCompanies(current_company_id, allowed_company_ids);
+        const toggleCompany = () => {
+            [...this._getElements()].forEach((item) => {
+                const companyID = parseInt(item.getAttribute('data-company-id'));
+                if (item.querySelector('.fa-check-square')) {
+                    if (!allowedCompanyIds.includes(companyID)) {
+                        allowedCompanyIds.push(companyID);
+                    }
+                } else if (allowedCompanyIds.includes(companyID)) {
+                    allowedCompanyIds.splice(allowedCompanyIds.indexOf(companyID), 1);
+                }
+            });
+            session.setCompanies(currentCompanyId, allowedCompanyIds);
+        };
+        if (this.toggleTimeout) {
+            clearTimeout(this.toggleTimeout);
+        }
+        this.toggleTimeout = setTimeout(toggleCompany, 1000);
```

Comment:
I'd set the timeout value (1000) as an attribute on the Widget, s.t. we can mock it in the tests (like we do for the autocomplete delay in Many2one for instance).